### PR TITLE
Improve performance on virtual method processing

### LINKED ISSUE #40060: PublishTrimmed build time regression.
image

Not happening with all applications. Reproduced on Windows and Linux.

Between SDKs (installer) 9.0.100-preview.2.24112.28 and 9.0.100-preview.2.24113.7

Before:
```
crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/plaintext.benchmarks.yml --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/build/ci.profile.yml --scenario plaintext --profile intel-win-app --profile intel-load2-load --application.buildArguments "/p:PublishReadyToRun=true /p:PublishSingleFile=true /p:PublishTrimmed=true" --application.framework net9.0 --application.aspNetCoreVersion 9.0.0-preview.2.24112.24 --application.runtimeVersion 9.0.0-preview.2.24112.16 --application.sdkVersion 9.0.100-preview.2.24112.28 --profile noload
```

After
```
crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/plaintext.benchmarks.yml --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/build/ci.profile.yml --scenario plaintext --profile intel-win-app --profile intel-load2-load --application.buildArguments "/p:PublishReadyToRun=true /p:PublishSingleFile=true /p:PublishTrimmed=true" --application.framework net9.0 --application.aspNetCoreVersion 9.0.0-preview.2.24112.24 --application.runtimeVersion 9.0.0-preview.2.24112.16 --application.sdkVersion 9.0.100-preview.2.24113.7 --profile noload
```

### PR DESCRIPTION
**Pull request by jtschuster**
Fixes https://github.com/dotnet/sdk/issues/40060.

The _virtual_methods cache had duplicates due to the `Scope` item in the tuple, which made ProcessVirtualMethods do significantly more work. This PR makes the cache a dictionary, with TryAdd's instead of Add to only add each method once. ~~It also exits early when possible in the IsInterfaceImplementationMarkedRecursively method.~~ These changes made caused the `overrides` cache in TypeMapInfo to be modified during iteration, so a `for` loop is used instead of a foreach loop. This should be find since the list will always be append-only. I noticed we copy _typesWithInterfaces to an array for the same reason, though since that's also append-only we can iterate with a `for` loop there, too.

On my machine, this reduces the trim time of the aspnetcore benchmarks from ~80 seconds to ~35 seconds.

### CODE REVIEW

**Review by sbomer — COMMENTED** [review]
> It also exits early when possible in the IsInterfaceImplementationMarkedRecursively method.

Looks like you reverted that change, but there are still some left-over related diffs.

**Review by sbomer — APPROVED** [review]
Good find, thank you!

File: src/tools/illink/src/linker/Linker.Steps/MarkStep.cs [hunk 1]
```diff
@@ -220,7 +220,7 @@ public partial class MarkStep : IStep
 		public MarkStep ()
 		{
 			_methods = new Queue<(MethodDefinition, DependencyInfo, MessageOrigin)> ();
-			_virtual_methods = new HashSet<(MethodDefinition, MarkScopeStack.Scope)> ();
+			_virtual_methods = new Dictionary<MethodDefinition, MarkScopeStack.Scope> ();
```
Comment (sebastienros) [inline src/tools/illink/src/linker/Linker.Steps/MarkStep.cs:223]: Is the perf improvement due to the change of key? If so is `MethodDefinition` actually a good option for lookups? It doesn't implement `IEquatable` and doesn't have a custom `GetHashcode`. Would a custom record using what makes a `MethodDefinition` identity better then?

Comment (jtschuster) [reply src/tools/illink/src/linker/Linker.Steps/MarkStep.cs:223]: The trimmer only creates a single `MethodDefinition` object for each method definition in metadata, so `object`'s implementation of `GetHashCode` and `Equals` are fine here.

**Comment by sebastienros** [comment]
Quick update on this PR, this definitely fixed the regression

image

**Pull request merged**